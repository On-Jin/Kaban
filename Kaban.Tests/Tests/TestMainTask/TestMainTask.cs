using System.Net;
using System.Text.Json;
using System.Text.Json.Nodes;
using FluentAssertions;
using Kaban.GraphQL.MainTasks;
using Kaban.Models.Dto;
using Kaban.Tests.Setup;
using Snapshooter;
using Snapshooter.Xunit;
using Xunit.Abstractions;

namespace Kaban.Tests.Tests.TestMainTask;

[Collection(nameof(WebAppCollectionFixture))]
public class TestMainTask(WebAppFactory factory, ITestOutputHelper testOutputHelper)
    : TestMainTaskBase(factory, testOutputHelper)
{
    [Fact]
    public async Task GraphQL_AddMainTask()
    {
        await ResetAndPopulateDb();

        {
            var response = await AddMainTask(HttpClientShadow);

            var jsonWrapped = await response.Content.ReadAsStringAsync();
            TestOutputHelper.WriteLine(jsonWrapped);

            JsonSerializer
                .Serialize(JsonSerializer.Deserialize<object>(jsonWrapped), TestHelper.JsonSerializerOptions)
                .MatchSnapshot();
        }
        PrettyJson(await GetQueryBoardsString()).MatchSnapshot(SnapshotNameExtension.Create("FullBoardQuery"));
    }


    [Fact]
    public async Task GraphQL_DeleteMainTask()
    {
        await ResetAndPopulateDb();

        var id = 0;
        {
            var response = await AddMainTask(HttpClientShadow);
            var jsonWrapped = await response.Content.ReadAsStringAsync();
            TestOutputHelper.WriteLine(jsonWrapped);
            var jsonObject = JsonNode.Parse(jsonWrapped)!.AsObject();
            id = int.Parse(jsonObject["data"]!["addMainTask"]!["mainTask"]!["id"]!.ToString());
            TestOutputHelper.WriteLine($"{id}");
        }

        {
            await DeleteMainTask(HttpClientShadow, new DeleteMainTaskInput(id));

            var jsonWrapped = await GetQueryBoardsString();
            var json = JsonNode.Parse(jsonWrapped)!["data"]!["boards"]!.ToString();
            var boardsDto = JsonSerializer.Deserialize<List<BoardDto>>(json,
                new JsonSerializerOptions() { PropertyNameCaseInsensitive = true });
            boardsDto![0].Columns[0]!.MainTasks.Should().NotContain(c => c.Id == id);
        }
    }


    [Fact]
    public async Task GraphQL_PatchMainTask()
    {
        await ResetAndPopulateDb();

        PrettyJson(await PatchMainTask(HttpClientShadow)).MatchSnapshot();
    }

    [Fact]
    public async Task GraphQL_MoveMainTask_DisplaceToOtherColumn()
    {
        await ResetAndPopulateDb();

        PrettyJson(await MoveMainTask(HttpClientShadow, new MoveMainTaskInput(1, "Doing", null)))
            .MatchSnapshot();
    }

    [Fact]
    public async Task GraphQL_MoveMainTask_WrongStatus()
    {
        await ResetAndPopulateDb();

        PrettyJson(await MoveMainTask(HttpClientShadow, new MoveMainTaskInput(1, "Wrong!", null),
                HttpStatusCode.InternalServerError))
            .MatchSnapshot();
    }

    [Fact]
    public async Task GraphQL_MoveMainTask_Order_Same_Negative()
    {
        await ResetAndPopulateForMainTasks();

        PrettyJson(await MoveMainTask(HttpClientShadow, new MoveMainTaskInput(1, null, -3),
                HttpStatusCode.InternalServerError))
            .MatchSnapshot();
    }

    [Fact]
    public async Task GraphQL_MoveMainTask_Order_Same_Over()
    {
        await ResetAndPopulateForMainTasks();

        PrettyJson(await MoveMainTask(HttpClientShadow, new MoveMainTaskInput(1, null, 4),
                HttpStatusCode.InternalServerError))
            .MatchSnapshot();
    }

    [Fact]
    public async Task GraphQL_MoveMainTask_Order_Same_Same()
    {
        await ResetAndPopulateForMainTasks();

        PrettyJson(await MoveMainTask(HttpClientShadow, new MoveMainTaskInput(1, null, 0)))
            .MatchSnapshot();
    }

    [Fact]
    public async Task GraphQL_MoveMainTask_Order_Same_0_3()
    {
        await ResetAndPopulateForMainTasks();

        PrettyJson(await MoveMainTask(HttpClientShadow, new MoveMainTaskInput(1, null, 3)))
            .MatchSnapshot();
    }

    [Fact]
    public async Task GraphQL_MoveMainTask_Order_Same_3_0()
    {
        await ResetAndPopulateForMainTasks();
    
        PrettyJson(await MoveMainTask(HttpClientShadow, new MoveMainTaskInput(4, null, 0)))
            .MatchSnapshot();
    }
    
    
    [Fact]
    public async Task GraphQL_MoveMainTask_Displace_Order_0()
    {
        await ResetAndPopulateForMainTasks();
    
        PrettyJson(await MoveMainTask(HttpClientShadow, new MoveMainTaskInput(3, "Column 1", 0)))
            .MatchSnapshot();
    }
    
    [Fact]
    public async Task GraphQL_MoveMainTask_Displace_Order_2()
    {
        await ResetAndPopulateForMainTasks();
    
        PrettyJson(await MoveMainTask(HttpClientShadow, new MoveMainTaskInput(3, "Column 1", 2)))
            .MatchSnapshot();
    }
    
    [Fact]
    public async Task GraphQL_MoveMainTask_Displace_Order_4()
    {
        await ResetAndPopulateForMainTasks();
    
        PrettyJson(await MoveMainTask(HttpClientShadow, new MoveMainTaskInput(3, "Column 1", 4)))
            .MatchSnapshot();
    }

    // [Fact]
    // public async Task GraphQL_MoveMainTask_DisplaceToOtherColumn_InsertTo_4()
}