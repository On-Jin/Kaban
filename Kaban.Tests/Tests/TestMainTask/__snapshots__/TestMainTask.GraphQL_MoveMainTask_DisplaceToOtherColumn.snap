{
  "data": {
    "moveMainTask": {
      "board": {
        "id": 1,
        "name": "Platform Launch",
        "columns": [
          {
            "id": 1,
            "name": "Todo",
            "mainTasks": [
              {
                "description": "",
                "id": 2,
                "status": "Todo",
                "title": "Build UI for search",
                "subTasks": [
                  {
                    "id": 4,
                    "isCompleted": false,
                    "title": "Search page"
                  }
                ]
              },
              {
                "description": "",
                "id": 3,
                "status": "Todo",
                "title": "Build settings UI",
                "subTasks": [
                  {
                    "id": 5,
                    "isCompleted": false,
                    "title": "Account page"
                  },
                  {
                    "id": 6,
                    "isCompleted": false,
                    "title": "Billing page"
                  }
                ]
              },
              {
                "description": "Once we feel version one is ready, we need to rigorously test it both internally and externally to identify any major gaps.",
                "id": 4,
                "status": "Todo",
                "title": "QA and test all major user journeys",
                "subTasks": [
                  {
                    "id": 7,
                    "isCompleted": false,
                    "title": "Internal testing"
                  },
                  {
                    "id": 8,
                    "isCompleted": false,
                    "title": "External testing"
                  }
                ]
              }
            ]
          },
          {
            "id": 2,
            "name": "Doing",
            "mainTasks": [
              {
                "description": "",
                "id": 5,
                "status": "Doing",
                "title": "Design settings and search pages",
                "subTasks": [
                  {
                    "id": 9,
                    "isCompleted": true,
                    "title": "Settings - Account page"
                  },
                  {
                    "id": 10,
                    "isCompleted": true,
                    "title": "Settings - Billing page"
                  },
                  {
                    "id": 11,
                    "isCompleted": false,
                    "title": "Search page"
                  }
                ]
              },
              {
                "description": "",
                "id": 6,
                "status": "Doing",
                "title": "Add account management endpoints",
                "subTasks": [
                  {
                    "id": 12,
                    "isCompleted": true,
                    "title": "Upgrade plan"
                  },
                  {
                    "id": 13,
                    "isCompleted": true,
                    "title": "Cancel plan"
                  },
                  {
                    "id": 14,
                    "isCompleted": false,
                    "title": "Update payment method"
                  }
                ]
              },
              {
                "description": "",
                "id": 7,
                "status": "Doing",
                "title": "Design onboarding flow",
                "subTasks": [
                  {
                    "id": 15,
                    "isCompleted": true,
                    "title": "Sign up page"
                  },
                  {
                    "id": 16,
                    "isCompleted": false,
                    "title": "Sign in page"
                  },
                  {
                    "id": 17,
                    "isCompleted": false,
                    "title": "Welcome page"
                  }
                ]
              },
              {
                "description": "",
                "id": 8,
                "status": "Doing",
                "title": "Add search enpoints",
                "subTasks": [
                  {
                    "id": 18,
                    "isCompleted": true,
                    "title": "Add search endpoint"
                  },
                  {
                    "id": 19,
                    "isCompleted": false,
                    "title": "Define search filters"
                  }
                ]
              },
              {
                "description": "",
                "id": 9,
                "status": "Doing",
                "title": "Add authentication endpoints",
                "subTasks": [
                  {
                    "id": 20,
                    "isCompleted": true,
                    "title": "Define user model"
                  },
                  {
                    "id": 21,
                    "isCompleted": false,
                    "title": "Add auth endpoints"
                  }
                ]
              },
              {
                "description": "We know what we\u0027re planning to build for version one. Now we need to finalise the first pricing model we\u0027ll use. Keep iterating the subtasks until we have a coherent proposition.",
                "id": 10,
                "status": "Doing",
                "title": "Research pricing points of various competitors and trial different business models",
                "subTasks": [
                  {
                    "id": 22,
                    "isCompleted": true,
                    "title": "Research competitor pricing and business models"
                  },
                  {
                    "id": 23,
                    "isCompleted": false,
                    "title": "Outline a business model that works for our solution"
                  },
                  {
                    "id": 24,
                    "isCompleted": false,
                    "title": "Talk to potential customers about our proposed solution and ask for fair price expectancy"
                  }
                ]
              },
              {
                "description": "",
                "id": 1,
                "status": "Doing",
                "title": "Build UI for onboarding flow",
                "subTasks": [
                  {
                    "id": 1,
                    "isCompleted": true,
                    "title": "Sign up page"
                  },
                  {
                    "id": 2,
                    "isCompleted": false,
                    "title": "Sign in page"
                  },
                  {
                    "id": 3,
                    "isCompleted": false,
                    "title": "Welcome page"
                  }
                ]
              }
            ]
          },
          {
            "id": 3,
            "name": "Done",
            "mainTasks": [
              {
                "description": "Ensure the layout continues to make sense and we have strong buy-in from potential users.",
                "id": 11,
                "status": "Done",
                "title": "Conduct 5 wireframe tests",
                "subTasks": [
                  {
                    "id": 25,
                    "isCompleted": true,
                    "title": "Complete 5 wireframe prototype tests"
                  }
                ]
              },
              {
                "description": "Create a greyscale clickable wireframe prototype to test our asssumptions so far.",
                "id": 12,
                "status": "Done",
                "title": "Create wireframe prototype",
                "subTasks": [
                  {
                    "id": 26,
                    "isCompleted": true,
                    "title": "Create clickable wireframe prototype in Balsamiq"
                  }
                ]
              },
              {
                "description": "Keep iterating through the subtasks until we\u0027re clear on the core concepts for the app.",
                "id": 13,
                "status": "Done",
                "title": "Review results of usability tests and iterate",
                "subTasks": [
                  {
                    "id": 27,
                    "isCompleted": true,
                    "title": "Meet to review notes from previous tests and plan changes"
                  },
                  {
                    "id": 28,
                    "isCompleted": true,
                    "title": "Make changes to paper prototypes"
                  },
                  {
                    "id": 29,
                    "isCompleted": true,
                    "title": "Conduct 5 usability tests"
                  }
                ]
              },
              {
                "description": "",
                "id": 14,
                "status": "Done",
                "title": "Create paper prototypes and conduct 10 usability tests with potential customers",
                "subTasks": [
                  {
                    "id": 30,
                    "isCompleted": true,
                    "title": "Create paper prototypes for version one"
                  },
                  {
                    "id": 31,
                    "isCompleted": true,
                    "title": "Complete 10 usability tests"
                  }
                ]
              },
              {
                "description": "We need to define and refine our core product. Interviews will help us learn common pain points and help us define the strongest MVP.",
                "id": 15,
                "status": "Done",
                "title": "Market discovery",
                "subTasks": [
                  {
                    "id": 32,
                    "isCompleted": true,
                    "title": "Interview 10 prospective customers"
                  }
                ]
              },
              {
                "description": "",
                "id": 16,
                "status": "Done",
                "title": "Competitor analysis",
                "subTasks": [
                  {
                    "id": 33,
                    "isCompleted": true,
                    "title": "Find direct and indirect competitors"
                  },
                  {
                    "id": 34,
                    "isCompleted": true,
                    "title": "SWOT analysis for each competitor"
                  }
                ]
              },
              {
                "description": "We need to get a solid overview of the market to ensure we have up-to-date estimates of market size and demand.",
                "id": 17,
                "status": "Done",
                "title": "Research the market",
                "subTasks": [
                  {
                    "id": 35,
                    "isCompleted": true,
                    "title": "Write up research analysis"
                  },
                  {
                    "id": 36,
                    "isCompleted": true,
                    "title": "Calculate TAM"
                  }
                ]
              }
            ]
          }
        ]
      }
    }
  }
}
